In [7]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
#import tensorflow as tf

import pandas as pd 
df = pd.read_excel("lung-data.xlsx")
df

FIPS_code    County State  Lung Cancer  PM2.5  Status Variable  \
0          1001   Autauga    AL         73.9  12.06                1   
1          1003   Baldwin    AL         68.4  11.12                1   
2          1005   Barbour    AL         76.1  12.36                1   
3          1007  Bibb Cou    AL         86.4  12.24                1   
4          1009  Blount C    AL         73.1  12.97                1   
...         ...       ...   ...          ...    ...              ...   
2597      56037  Sweetwat    WY         45.3   6.40                0   
2598      56039  Teton Co    WY         22.6   6.67                0   
2599      56041  Uinta Co    WY         37.1   6.57                0   
2600      56043  Washakie    WY         51.1   6.33                0   
2601      56045  Weston C    WY         46.4   5.24                0   

      Land_EQI  Sociod_EQI  Built_EQI    LTD  ...  Status   LCI    UCI  Inter  \
0    -0.706591    0.670436  -0.497301   4.79  ...       2  64.3   84.6  42.94   
1    -1.084299    0.553073   0.401585   3.68  ...       2  63.9   73.1  40.54   
2    -1.281470   -1.236294   0.048854   0.87  ...       2  63.3   90.9  77.48   
3    -0.827410   -0.600018  -1.290857  19.95  ...       2  71.2  104.1  -6.00   
4    -0.622934    0.296509  -1.262740  19.95  ...       2  64.5   82.6  37.83   
...        ...         ...        ...    ...  ...     ...   ...    ...    ...   
2597  0.329106    0.832869   1.309324   5.31  ...       1  35.4   57.0    NaN   
2598  0.316508    1.812128   2.418939   1.14  ...       1  14.1   34.4    NaN   
2599  0.253324    0.715908   1.134188  10.86  ...       1  25.2   52.5    NaN   
2600  0.151024   -0.067995   0.728454  10.52  ...       1  35.1   73.3    NaN   
2601  0.297553    0.043923   0.637936  10.52  ...       1  29.5   70.8    NaN   

       Slp  control    treat  Local_Treat  AAC      RT  
0     2.60  68.6143  73.4043       4.7900   44  stable  
1     2.76  66.0568  69.7357       3.6789  181  stable  
2    -0.11  75.7286  76.6000       0.8714   26  stable  
3     7.64  65.9182  85.8692      19.9510   23  stable  
4     3.36  69.9500  76.9564       7.0064   54  stable  
...    ...      ...      ...          ...  ...     ...  
2597   NaN      NaN      NaN          NaN   16  stable  
2598   NaN      NaN      NaN          NaN    5  stable  
2599   NaN      NaN      NaN          NaN    7  stable  
2600   NaN      NaN      NaN          NaN    7  stable  
2601   NaN      NaN      NaN          NaN    5  stable  

[2602 rows x 34 columns]

In [8]:
df.columns

Index(['FIPS_code', 'County', 'State', 'Lung Cancer', 'PM2.5',
       'Status Variable', 'Land_EQI', 'Sociod_EQI', 'Built_EQI', 'LTD',
       'Intercept', 'Slope', 'CLU50_1', 'PM10', 'SO2', 'NO2', 'O3', 'CO', 'CN',
       'Disel', 'CS2', 'Air_EQI', 'Water_EQI', 'EQI', 'Status', 'LCI', 'UCI',
       'Inter', 'Slp', 'control', 'treat', 'Local_Treat', 'AAC', 'RT'],
      dtype='object')

In [9]:
df_preprocessed = df.drop(columns=['County','State','Status Variable','LTD','Intercept','Slope','Status','Inter',
                                   'Slp','control','treat','Local_Treat','Lung Cancer'])
df_preprocessed

FIPS_code  PM2.5  Land_EQI  Sociod_EQI  Built_EQI  CLU50_1   PM10  \
0          1001  12.06 -0.706591    0.670436  -0.497301        1  15.07   
1          1003  11.12 -1.084299    0.553073   0.401585        2  19.99   
2          1005  12.36 -1.281470   -1.236294   0.048854        3  15.77   
3          1007  12.24 -0.827410   -0.600018  -1.290857        4  14.92   
4          1009  12.97 -0.622934    0.296509  -1.262740        5  17.90   
...         ...    ...       ...         ...        ...      ...    ...   
2597      56037   6.40  0.329106    0.832869   1.309324       40  11.26   
2598      56039   6.67  0.316508    1.812128   2.418939       47  10.40   
2599      56041   6.57  0.253324    0.715908   1.134188       39   8.35   
2600      56043   6.33  0.151024   -0.067995   0.728454       13   7.21   
2601      56045   5.24  0.297553    0.043923   0.637936       13   4.16   

              SO2         NO2       O3  ...        CN     Disel       CS2  \
0       10.661088  123.657648   522.38  ...  0.054815  0.388556  0.008080   
1       17.146847  247.742253   540.79  ...  0.021069  0.428278  0.001090   
2       23.257118  183.193624   896.42  ...  0.014027  0.199725  0.000513   
3        7.630953  127.779935   563.48  ...  0.009613  0.211741  0.000225   
4        8.913795   95.198094   561.94  ...  0.022128  0.300100  0.000429   
...           ...         ...      ...  ...       ...       ...       ...   
2597    51.633922  342.108999  1703.31  ...  0.007201  0.204528  0.000227   
2598    39.036416  322.610755  3614.23  ...  0.005966  0.080880  0.000114   
2599    65.490741  529.309963  3467.05  ...  0.002770  0.089568  0.000245   
2600   157.728659  284.452099  2819.26  ...  0.011684  0.157553  0.000184   
2601  1207.253574  488.002519  5283.84  ...  0.002027  0.059740  0.000039   

       Air_EQI  Water_EQI   EQI   LCI    UCI  AAC      RT  
0     0.955385  -1.109728  0.00  64.3   84.6   44  stable  
1     0.717964  -0.565911  0.20  63.9   73.1  181  stable  
2     0.131007  -0.978090 -0.95  63.3   90.9   26  stable  
3     0.065289  -0.968173 -1.09  71.2  104.1   23  stable  
4     0.402194  -0.718645 -0.51  64.5   82.6   54  stable  
...        ...        ...   ...   ...    ...  ...     ...  
2597 -0.039545   1.130375  1.09  35.4   57.0   16  stable  
2598 -0.557297   1.071576  1.67  14.1   34.4    5  stable  
2599 -0.814369   1.162652  0.69  25.2   52.5    7  stable  
2600 -0.244298   1.004629  0.38  35.1   73.3    7  stable  
2601 -1.638539   1.039474 -0.04  29.5   70.8    5  stable  

[2602 rows x 21 columns]

In [10]:
df_preprocessed.dtypes

FIPS_code       int64
PM2.5         float64
Land_EQI      float64
Sociod_EQI    float64
Built_EQI     float64
CLU50_1         int64
PM10          float64
SO2           float64
NO2           float64
O3            float64
CO            float64
CN            float64
Disel         float64
CS2           float64
Air_EQI       float64
Water_EQI     float64
EQI           float64
LCI           float64
UCI           float64
AAC             int64
RT             object
dtype: object

In [11]:
df_preprocessed = df_preprocessed.dropna()

In [12]:
df_preprocessed_1 = df_preprocessed[df_preprocessed.RT != '*']
df_preprocessed_1

FIPS_code  PM2.5  Land_EQI  Sociod_EQI  Built_EQI  CLU50_1   PM10  \
0          1001  12.06 -0.706591    0.670436  -0.497301        1  15.07   
1          1003  11.12 -1.084299    0.553073   0.401585        2  19.99   
2          1005  12.36 -1.281470   -1.236294   0.048854        3  15.77   
3          1007  12.24 -0.827410   -0.600018  -1.290857        4  14.92   
4          1009  12.97 -0.622934    0.296509  -1.262740        5  17.90   
...         ...    ...       ...         ...        ...      ...    ...   
2597      56037   6.40  0.329106    0.832869   1.309324       40  11.26   
2598      56039   6.67  0.316508    1.812128   2.418939       47  10.40   
2599      56041   6.57  0.253324    0.715908   1.134188       39   8.35   
2600      56043   6.33  0.151024   -0.067995   0.728454       13   7.21   
2601      56045   5.24  0.297553    0.043923   0.637936       13   4.16   

              SO2         NO2       O3  ...        CN     Disel       CS2  \
0       10.661088  123.657648   522.38  ...  0.054815  0.388556  0.008080   
1       17.146847  247.742253   540.79  ...  0.021069  0.428278  0.001090   
2       23.257118  183.193624   896.42  ...  0.014027  0.199725  0.000513   
3        7.630953  127.779935   563.48  ...  0.009613  0.211741  0.000225   
4        8.913795   95.198094   561.94  ...  0.022128  0.300100  0.000429   
...           ...         ...      ...  ...       ...       ...       ...   
2597    51.633922  342.108999  1703.31  ...  0.007201  0.204528  0.000227   
2598    39.036416  322.610755  3614.23  ...  0.005966  0.080880  0.000114   
2599    65.490741  529.309963  3467.05  ...  0.002770  0.089568  0.000245   
2600   157.728659  284.452099  2819.26  ...  0.011684  0.157553  0.000184   
2601  1207.253574  488.002519  5283.84  ...  0.002027  0.059740  0.000039   

       Air_EQI  Water_EQI   EQI   LCI    UCI  AAC      RT  
0     0.955385  -1.109728  0.00  64.3   84.6   44  stable  
1     0.717964  -0.565911  0.20  63.9   73.1  181  stable  
2     0.131007  -0.978090 -0.95  63.3   90.9   26  stable  
3     0.065289  -0.968173 -1.09  71.2  104.1   23  stable  
4     0.402194  -0.718645 -0.51  64.5   82.6   54  stable  
...        ...        ...   ...   ...    ...  ...     ...  
2597 -0.039545   1.130375  1.09  35.4   57.0   16  stable  
2598 -0.557297   1.071576  1.67  14.1   34.4    5  stable  
2599 -0.814369   1.162652  0.69  25.2   52.5    7  stable  
2600 -0.244298   1.004629  0.38  35.1   73.3    7  stable  
2601 -1.638539   1.039474 -0.04  29.5   70.8    5  stable  

[2555 rows x 21 columns]

In [13]:
df_preprocessed_1['RT'].value_counts()

stable     2267
falling     237
rising       51
Name: RT, dtype: int64

In [14]:
df_preprocessed_1['RT'] = df_preprocessed_1['RT'].replace({'stable': 0, 'rising': 1, 'falling':2})

/var/folders/qh/44677dhj5kv978487x28nvmm0000gn/T/ipykernel_12616/14303869.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_preprocessed_1['RT'] = df_preprocessed_1['RT'].replace({'stable': 0, 'rising': 1, 'falling':2})


In [15]:
df_preprocessed_1['RT'] = df_preprocessed_1['RT'].astype(int)

/var/folders/qh/44677dhj5kv978487x28nvmm0000gn/T/ipykernel_12616/7156509.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_preprocessed_1['RT'] = df_preprocessed_1['RT'].astype(int)


In [16]:
# Note: this classification task is not directly related to project 4. I was simply checking how different classifier 
# can predict the air quality class ('stable', 'falling' and 'rising') from RT column based on different input features
# Split our preprocessed data into our features and target arrays
y = df_preprocessed_1['RT'].values
X = df_preprocessed_1.drop('RT', axis=1).values
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.20,random_state = 42)

In [17]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier()

# Train Decision Tree Classifer
clf = clf.fit(X_train,y_train)
y_pred = clf.predict(X_train)


In [18]:
from sklearn import metrics
# Model Accuracy on train dataset, how often is the classifier correct?
print("Accuracy:", metrics.accuracy_score(y_train, y_pred))

Accuracy: 1.0


In [19]:
# Model Accuracy on test dataset
y_pred1 = clf.predict(X_test)
print("Accuracy:",metrics.accuracy_score(y_test, y_pred1))

Accuracy: 0.8414872798434442


In [20]:
confusion_matrix = metrics.confusion_matrix(y_test, y_pred1)
confusion_matrix

array([[411,   8,  32],
       [  7,   0,   0],
       [ 30,   4,  19]])

In [21]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred1))

              precision    recall  f1-score   support

           0       0.92      0.91      0.91       451
           1       0.00      0.00      0.00         7
           2       0.37      0.36      0.37        53

    accuracy                           0.84       511
   macro avg       0.43      0.42      0.43       511
weighted avg       0.85      0.84      0.84       511



In [22]:
from sklearn.svm import SVC
from sklearn.pipeline import make_pipeline
clf_svm = make_pipeline(StandardScaler(), SVC(gamma='auto'))
clf_svm.fit(X_train, y_train)
y_pred_svm = clf_svm.predict(X_test)

In [23]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred_svm))

Accuracy: 0.8923679060665362


In [24]:
confusion_matrix_svm = metrics.confusion_matrix(y_test, y_pred_svm)
confusion_matrix_svm

array([[449,   0,   2],
       [  7,   0,   0],
       [ 46,   0,   7]])

In [25]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred_svm))

              precision    recall  f1-score   support

           0       0.89      1.00      0.94       451
           1       0.00      0.00      0.00         7
           2       0.78      0.13      0.23        53

    accuracy                           0.89       511
   macro avg       0.56      0.38      0.39       511
weighted avg       0.87      0.89      0.86       511



/Users/indranilroy/opt/anaconda3/envs/PythonData38/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1327: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/indranilroy/opt/anaconda3/envs/PythonData38/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1327: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/indranilroy/opt/anaconda3/envs/PythonData38/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1327: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  

In [26]:
# Note: All regression tasks are related to Project 4
df_regr = df.drop(columns=['FIPS_code','County','State','Status Variable','LTD','Intercept','Slope','Status','Inter',
                           'Slp','control','treat','Local_Treat','AAC','EQI','RT'])
df_regr

Lung Cancer  PM2.5  Land_EQI  Sociod_EQI  Built_EQI  CLU50_1   PM10  \
0            73.9  12.06 -0.706591    0.670436  -0.497301        1  15.07   
1            68.4  11.12 -1.084299    0.553073   0.401585        2  19.99   
2            76.1  12.36 -1.281470   -1.236294   0.048854        3  15.77   
3            86.4  12.24 -0.827410   -0.600018  -1.290857        4  14.92   
4            73.1  12.97 -0.622934    0.296509  -1.262740        5  17.90   
...           ...    ...       ...         ...        ...      ...    ...   
2597         45.3   6.40  0.329106    0.832869   1.309324       40  11.26   
2598         22.6   6.67  0.316508    1.812128   2.418939       47  10.40   
2599         37.1   6.57  0.253324    0.715908   1.134188       39   8.35   
2600         51.1   6.33  0.151024   -0.067995   0.728454       13   7.21   
2601         46.4   5.24  0.297553    0.043923   0.637936       13   4.16   

              SO2         NO2       O3           CO        CN     Disel  \
0       10.661088  123.657648   522.38     4.463225  0.054815  0.388556   
1       17.146847  247.742253   540.79    12.875833  0.021069  0.428278   
2       23.257118  183.193624   896.42    19.620539  0.014027  0.199725   
3        7.630953  127.779935   563.48     2.951976  0.009613  0.211741   
4        8.913795   95.198094   561.94     9.362215  0.022128  0.300100   
...           ...         ...      ...          ...       ...       ...   
2597    51.633922  342.108999  1703.31   422.322102  0.007201  0.204528   
2598    39.036416  322.610755  3614.23  3754.236317  0.005966  0.080880   
2599    65.490741  529.309963  3467.05  1109.005206  0.002770  0.089568   
2600   157.728659  284.452099  2819.26   199.103987  0.011684  0.157553   
2601  1207.253574  488.002519  5283.84    25.799703  0.002027  0.059740   

           CS2   Air_EQI  Water_EQI   LCI    UCI  
0     0.008080  0.955385  -1.109728  64.3   84.6  
1     0.001090  0.717964  -0.565911  63.9   73.1  
2     0.000513  0.131007  -0.978090  63.3   90.9  
3     0.000225  0.065289  -0.968173  71.2  104.1  
4     0.000429  0.402194  -0.718645  64.5   82.6  
...        ...       ...        ...   ...    ...  
2597  0.000227 -0.039545   1.130375  35.4   57.0  
2598  0.000114 -0.557297   1.071576  14.1   34.4  
2599  0.000245 -0.814369   1.162652  25.2   52.5  
2600  0.000184 -0.244298   1.004629  35.1   73.3  
2601  0.000039 -1.638539   1.039474  29.5   70.8  

[2602 rows x 18 columns]

In [28]:
from sklearn.ensemble import RandomForestRegressor
# Split our preprocessed data into our features and target arrays
y = df_regr['Lung Cancer'].values
X = df_regr.drop('Lung Cancer', axis=1).values
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state = 42)
model_regr = RandomForestRegressor(max_depth=6, random_state=42)
model_regr.fit(X_train, y_train)

RandomForestRegressor(max_depth=6, random_state=42)

In [29]:
model_regr.feature_importances_

array([6.24690692e-05, 1.50613015e-04, 2.12625889e-04, 1.75602384e-04,
       2.60978732e-05, 9.09408791e-05, 3.85341505e-05, 1.19881046e-04,
       4.79536952e-05, 6.21677750e-05, 1.07480916e-04, 1.33023713e-04,
       9.94938674e-05, 8.20169047e-05, 1.37933646e-04, 6.53062239e-01,
       3.45390926e-01])

In [106]:
# feature_names = [f"feature {i}" for i in range(X.shape[1])]

In [107]:
# import pandas as pd
# import matplotlib.pyplot as plt
# plt.barh(feature_names, model_regr.feature_importances_)

In [33]:
y_pred_rf = model_regr.predict(X_test)

Quantifying RandomForestRegressor model with metrices below 

mean squared error (MSE) and
R2 Score

Note: There are a variety of ways to quantify the model, but MSE and R2 are quite common

In [79]:
from sklearn.metrics import mean_squared_error, r2_score

# Score the prediction with MSE and R2
mse_rf = mean_squared_error(y_test, y_pred_rf, squared=False)
r2_rf = r2_score(y_test, y_pred_rf)

print(f"mean squared error (MSE): {mse_rf}")
print(f"R-squared (R2 ): {r2_rf}")

mean squared error (MSE): 1.0901631707664408
R-squared (R2 ): 0.9961216810575816


A good MSE score will be close to zero, while a good R2 score will be close to 1.
The obtained value suggest that RandomForestRegressor works efficiently on the dataset

In [67]:
# Five-fold cross validation
from sklearn.model_selection import cross_val_score
cv_results_regr = cross_val_score(model_regr, X, y, cv=5, scoring='r2')

In [68]:
cv_results_regr

array([0.99646495, 0.93781531, 0.99576459, 0.99580746, 0.99470601])

Five-fold cross validation results suggest RandomForestRegressor works efficiently on the dataset 

In [69]:
from sklearn import model_selection
import pickle

filename = 'finalizedregr_model.pt'
pickle.dump(model_regr, open(filename, 'wb'))
 
# some time later...
 
# load the model from disk
loaded_model = pickle.load(open(filename, 'rb'))
result = loaded_model.predict(X_test)
print(result)

[ 57.7648638   46.85547781  33.40084775  62.51101757  74.89374967
  52.56088833  74.56877779  70.98854381  70.58827529  45.17823939
  84.47462799  56.06408139  49.55325293  62.66308084  83.37399699
  52.56819649  81.44958104  88.92597903  75.46203028  81.97281783
  80.22743672  91.03021485  81.65327163  61.12095647  70.34608887
  38.01265699  97.98824443  44.25443048  73.32251437  82.17956157
  64.38676725  51.86316789  52.91180204  59.33100336  79.68943288
  81.34866192  39.50503983  42.55825702  58.84918404  54.86640206
  59.56975591  48.33246414  62.40805899  84.68621307  57.93883936
  93.86825674  45.42573624  71.3661084   64.85189539  59.39882848
  87.31805782  41.26095859 118.05681793  44.7659286   49.5577861
  66.32996744  53.84994187  62.99821012  74.74963659  55.13102702
  47.85523961  41.60728665  61.85545283  52.35083657  75.7690999
  66.63439289  66.86077898  64.97036714  86.22324251  46.07445108
  62.09832956  65.85635507  62.80264537  82.60714951  77.06121925
  74.2733296

In [70]:
# Checking how accurate predicted value is and compared it with ground truth
result2 = loaded_model.predict(X_test[2].reshape(1, -1))
print(result2)

[33.40084775]


In [71]:
y_test[2]

33.7

In [72]:
Input = X_test[2].reshape(1, -1)
Input
# len(Input) 

array([[ 6.06000000e+00,  8.01560300e-01,  1.53860500e-01,
         1.49720300e+00,  4.50000000e+01,  1.95700000e+01,
         1.20890739e+01,  2.90948333e+02,  1.60998000e+03,
         1.65366222e+02,  4.78040000e-03,  7.82166000e-02,
         1.54000000e-04, -1.37494800e+00,  8.82228400e-01,
         2.03000000e+01,  5.23000000e+01]])

In [73]:
print(df_regr.drop('Lung Cancer', axis=1). columns)  

Index(['PM2.5', 'Land_EQI', 'Sociod_EQI', 'Built_EQI', 'CLU50_1', 'PM10',
       'SO2', 'NO2', 'O3', 'CO', 'CN', 'Disel', 'CS2', 'Air_EQI', 'Water_EQI',
       'LCI', 'UCI'],
      dtype='object')


User inputs array in the interactive html page should have all these inputs from user to obtain lung
cancer mortality 

In [74]:
from sklearn.svm import SVR
svm_regr = make_pipeline(StandardScaler(), SVR(C=3, epsilon=0.1))
svm_regr.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()), ('svr', SVR(C=3))])

In [75]:
y_pred_svm = svm_regr.predict(X_test)

Quantifying SVR model with metrices below
mean squared error (MSE) and R2 Score

In [78]:
# Score the prediction with MSE and R2
mse_SVR = mean_squared_error(y_test, y_pred_svm, squared=False)
r2_SVR = r2_score(y_test, y_pred_svm)

print(f"mean squared error (MSE): {mse_SVR}")
print(f"R-squared (R2 ): {r2_SVR}")

mean squared error (MSE): 3.286872270493343
R-squared (R2 ): 0.9647445579626635


Results demonstrate the fact that SVR model is less accurate than 
RandomForestRegressor model

In [89]:
filename = 'finalizedsvm_model.pt'
pickle.dump(svm_regr, open(filename, 'wb'))
 
# some time later...
 
# load the model from disk
loaded_model = pickle.load(open(filename, 'rb'))
result = loaded_model.predict(X_test)
print(result)

[ 56.91122489  50.0281426   38.66788425  61.80732779  74.14843894
  52.71781388  75.10287643  71.69175428  70.73752391  47.11946336
  85.49714931  54.803725    60.17676999  62.14700098  83.69825203
  53.5206726   80.97258863  90.55655596  75.96658257  82.76096898
  80.36823827  91.78034863  77.80073197  59.51996882  72.80268354
  38.81389714 100.69976607  49.55150353  73.46008251  82.11931858
  65.38149859  53.11236883  51.73641159  58.29159244  80.33599509
  79.50470177  39.5029114   54.87950845  59.02469161  57.23790303
  59.63386778  47.83632115  61.85562878  84.14243646  57.67320027
  92.57528101  44.89281077  70.48084868  64.91635941  60.20403265
  87.95194853  41.91911942 113.61219786  43.07733122  47.1995073
  66.7035903   54.15032763  64.07600092  73.58896178  54.0435463
  48.24258428  46.94830581  61.14448346  52.71128489  76.4179061
  67.82720594  64.59543451  64.81297426  88.10777588  50.719863
  60.54767481  65.70137446  62.66716492  83.62598784  76.38836165
  74.87665376  

In [36]:
# Checking how accurate predicted value is and compared it with ground truth
result2 = loaded_model.predict(X_test[2].reshape(1, -1))
print(result2)

[38.66788425]


In [85]:
# Five-fold cross validation
cv_results_regr_svm = cross_val_score(svm_regr, X, y, cv=5, scoring='r2')

In [86]:
cv_results_regr_svm

array([0.91820326, 0.81007791, 0.96688675, 0.86465052, 0.86909039])

Five-fold cross validation results suggest SVR works efficiently on the dataset 

In [82]:
#Testing linear regression model
from sklearn.linear_model import LinearRegression
linear_regr = LinearRegression()
linear_regr.fit(X_train, y_train)

LinearRegression()

In [83]:
y_pred_linear = linear_regr.predict(X_test)

In [84]:
# Score the prediction with MSE and R2
mse_linear = mean_squared_error(y_test, y_pred_linear, squared=False)
r2_linear = r2_score(y_test, y_pred_linear)

print(f"mean squared error (MSE): {mse_linear}")
print(f"R-squared (R2 ): {r2_linear}")

mean squared error (MSE): 0.5137256106065202
R-squared (R2 ): 0.9991387624768184


Results demonstrate the fact that LinearRegression performs much better than SVR model and its metricess are a 
little better than RandomForestRegressor model

In [87]:
# Five-fold cross validation
cv_results_regr_linear = cross_val_score(linear_regr, X, y, cv=5, scoring='r2')

In [88]:
cv_results_regr_linear

array([0.99918381, 0.99934678, 0.99863768, 0.99670138, 0.99907259])

Five-fold cross validation results suggest LinearRegression works efficiently on the dataset 

In [90]:
filename = 'finalizedlinear_model.pt'
pickle.dump(linear_regr, open(filename, 'wb'))
 
# some time later...
 
# load the model from disk
loaded_model = pickle.load(open(filename, 'rb'))
result = loaded_model.predict(X_test)
print(result)

[ 58.50002876  47.18006099  33.18680215  61.6151303   75.31599853
  51.87136171  75.27547307  71.84029164  69.70301997  44.60645996
  84.03908514  56.42504058  49.49084598  62.39635635  83.84759097
  53.09356182  81.33099808  89.45844529  77.5474407   82.12013573
  80.96947675  90.94511873  81.98816863  59.84262205  72.4958778
  38.57812457  99.47570497  44.31810191  72.8643532   81.66391431
  64.80236041  51.65135729  51.34916748  59.11093437  80.22351398
  80.97114676  40.87261981  40.73617449  59.17987269  54.52703951
  59.81384494  48.42647281  62.58311644  86.09635459  57.19433843
  93.24000924  43.93066088  73.11877427  64.82763941  59.66471468
  87.09773273  40.94711439 119.87026464  44.6035269   48.5812553
  66.89454429  53.54071947  63.83834328  75.26753348  56.02595316
  47.89750538  38.88574149  61.70352047  53.41914423  76.47118849
  67.76755016  64.36898416  65.01832764  87.17866189  45.58591195
  61.41314134  66.08510362  62.6831312   83.31458748  76.15524552
  74.5580027

# FINAL ANALYSES:
Model performances: LinearRegression > RandomForestRegressor >> SVR model (based on MSE and r2 score)
LinearRegression model works a touch better than RandomForestRegressor model.
Both LinearRegression and RandomForestRegressor works much better than SVR model.
However, with MSE = 3.286872270493343 and r2 = 0.9647445579626635, SVR model can also be used for prediction for this particular dataset. 

We think, overall, our findings can help understanding the use of various supervised learning algorithms for predicting lung cancer mortality based on various air pollutants features as well as land and water quality indexes. However, one must understand the various limitations of this work before considering any predicted mortality output for practical use. The results solely depend on the dataset, data range used for training, and the features considered for training.    